### DISCUSSION TITLE
Is there way to Create Database and Tables on startup?

### DISCUSSION THREAD

**Discussion by MohsinDev369** [honojs/hono#2437] [Q&A] [answered]
Maybe a better title for this question, "How to run a piece of code on startup only"
Something like this, 
```
@app.on_event("startup")
def on_startup():
    create_db_and_tables()
```
The above code is from using Fastapi But I want to do this in hono.

Simply put, I want to run some code that should be called only once at startup, not before every request, maybe I can use singleton pattern but is there a built-in method like FastApi has?

**Comment by marcosrjjunior** [comment]
hey Mohsin,
could you explain a bit more your use case, and what database are you using?

usually you don't need to perform a "create" every time, because if is already created why would you want to perform the same query again?

One way of controlling that is by using [migrations](https://kysely.dev/docs/migrations).

For instance, using kysqly to connect to your DB:
- you can create a migration file like [this one](https://github.com/marcosrjjunior/hono-boilerplate/tree/main/src/lib/db/migrations).
- Then you can run [pnpm db:migrate:up](https://github.com/marcosrjjunior/hono-boilerplate/blob/main/package.json#L13). This will perform the queries under the `up` function.

This command is usually called on every deployment to apply the new required changes, then you can start your code as usual.

---

If your question is not related to the DB, and you wish to perform a code once on startup, you can add a new line after the new hono instance
```ts
const app = new Hono()

app.onError((error, c) => {  
  // ... 
})

// ...
console.log('initial load')
```

**Reply by MohsinDev369** [reply]
Hi, 
Thanks for the replay, appreciate it. So basically, I want to "create a table (for example todo table) if it does not exist" Programmatically, in fastAPI we have a "startup" event that we can use to run a piece of code when the app first starts then it will never run again a perfect place for code like creating table,  for more info on this you can read [this](https://sqlmodel.tiangolo.com/tutorial/fastapi/simple-hero-api/#create-database-and-tables-on-startup) which I am trying to replicate in hono.

All the table creation will be done through the Hono server, with no manual tempering with the db dashboard. Then all I need to do is give a connection string and run the server and tables will be created if it does not, For example, I have 100 tables that will be created programmatically, I don't want to do that manually,

**Reply by marcosrjjunior** [reply]
I'm not familiar with that approach.

You can easily generate migration files containing your table changes and run the cli to perform the changes without even restarting your server if that is the case.

**Comment by NicoPlyley** [comment, marked as answer]
Hono is only a router not the server, so it does not have control of those lifecycles methods like python frameworks do. This would depend on the runtime. For example in Bun you can import the Hono instance and put it in a fetch. From there you can access the startup and exit and run functions before each. You can even control if this would rerun every time the code is hot reloaded.
